foursight_core/app_utils.py

```diff
@@ -1817,6 +1849,17 @@ def collect_run_info(cls, run_uuid):
         # eliminate duplicates
         return set(complete)
 
+    _singleton = None
+    @staticmethod
+    def singleton(cls = None):
+        # A little wonky having a singleton with an argument but this is sort of the way it
+        # was in 4dn-cloud-infra/app-{cgap,fourfront}.py; and we know the only place we create
+        # this is from there and also from foursight-cgap/chalicelib/app.py and foursight/app.py
+        # with the appropriate locally derived (from this AppUtilsCore) AppUtils.
+        if not AppUtilsCore._singleton:
+            AppUtilsCore._singleton = cls() if cls else AppUtilsCore()
+        return AppUtilsCore._singleton
+
 
```

Comment:
Please see the `Singleton` class pending in [dcicutils PR 225](https://github.com/4dn-dcic/utils/pull/225) for `dcicutils.misc_utils` in dcicutils beta version `6.0.0.1b0` and code in the pending PR [here](https://github.com/4dn-dcic/utils/pull/225/files#diff-7e014960b5c3e0a844c1c60ee8ba98ca1c45f4f9b41ede991f1c6f67a43a49ebR1638).

It should look like:
```
from dcicutils.misc_utils import Singleton

class AppUtilsCore(Singleton, ...):
    ...   
```
and that will be enough to make a `.singleton` that returns `cls()` for the exact class of the named class. e.g., for `AppUtilsCore.singleton` you'll get a cached `AppUtilsCore()` and if you do a subclass Foo you'll get a cached `Foo()`.

If you use that class, the intent is that you should add `Singleton` to the classes mixed into AppUtilsCore and then expect AppUtilsCore and any of its subclasses to behave as if they have a `.singleton` property that would be computed for each such subclass. 

If this protocol you're using here is already in use in some way I don't see (it looks like you're adding a new method), it would be an incompatible change, but a big improvement, I think. 

Alternatively, you could implement the behavior you've defined here more cleanly by the following, but I really personally don't like the thing where you're passing a class to what should know its own methods. The intended thing is just that you let the class know how to make its own singleton. So instead of `AppUtilsCore.singleton(AppUtils)`, you want `AppUtils.singleton()` or `AppUtils.singleton` if you use my new class as a mixin to `AppUtilsCore`.
```
from dcicutils.misc_utils import check_true, classproperty_cached

@classproperty_cached
def _singleton(cls):
    return cls()

@staticmethod
def singleton(cls = None):
    check_true(issubclass(cls, AppUtilsCore),
               "The argument to .singleton() must be a subclass of AppUtilsCore.")
    return cls._singleton
```

Reply:
Note that if this utils beta is branched off of master you will receive incompatible ES7 changes - might need to do this later after my changes go in.